### PR TITLE
Issue/91. Handle patch correctly.

### PR DESCRIPTION
**Pull request by slash3b**
**TLDR**: priority fix and added some more tests   

Fixed major flaw — turned out  within extent of same version `patch` flag is of the highest priority.  
Priority sequence should be the following:   
```
1.1-alpha < 1.1-beta < 1.1-rc < 1.1-stable < 1.1 < 1.1-p
```
diff between old version and fixed one: 
```
slash3b@Nostromo ~/P/SecurityAdvisoriesBuilder (issue/91)> diff -u old_composer.json build/composer.json
--- old_composer.json   2019-12-13 18:25:18.956262258 +0200
+++ build/composer.json 2019-12-13 18:27:19.958957880 +0200
@@ -75,7 +75,7 @@
         "league/commonmark": "<0.18.3",
         "magento/magento1ce": "<1.9.4.3",
         "magento/magento1ee": ">=1,<1.14.4.3",
-        "magento/product-community-edition": ">=2,<2.2.10|>=2.3,<2.3.2",
+        "magento/product-community-edition": ">=2,<2.2.10|>=2.3,<2.3.2-p.2",
         "monolog/monolog": ">=1.8,<1.12",
         "namshi/jose": "<2.2",
         "onelogin/php-saml": "<2.10.4",
```

### CODE REVIEW

File: test/RoaveTest/SecurityAdvisories/VersionTest.php [hunk 1]
```diff
@@ -221,6 +221,27 @@ public function greaterThanComparisonVersionsProvider() : array
...
+            ['2.3.2-p2', '2.3.2-alpha1', true, false],
+            // compare with version higher
+            ['2.3.2-p2', '2.3.3', false, true],
+            // compare with version lower
+            ['2.3.2-p2', '2.3.1', true, false],
+            // compare two patches but with additional stability versions like 1-p1 and 1-p2 so the p2 will be greater
+            ['2.3.2-p1', '2.3.2-p2', false, true],
+            ['2.3.2-p2', '2.3.2-p2', false, false],
+            // @stof example
+            // 2.3.2-alpha1 < 2.3.2-beta1 < 2.3.2-rc1 < 2.3.2 < 2.3.2-p2 < 2.3.3-alpha1
```
Comment (Ocramius) [inline test/RoaveTest/SecurityAdvisories/VersionTest.php]: Commented out on purpose?

File: src/Roave/SecurityAdvisories/Version.php [hunk 2]
```diff
@@ -195,8 +187,45 @@ private static function removeTrailingZeroes(int ...$versionNumbers) : array
...
     }
 
+    /**
+     * Bear in mind that we compare flags only when versions are equal
+     *
+     * @param Version $other
+     */
     private function compareFlags(self $other) : int
     {
+        $patchLiterals = [
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/Version.php]: This is... Kinda terrible 😅

Still, good work on isolating it!

**Comment by nikolaposa** [comment]
Thanks for the ping @Ocramius. 👍 

On a different note, these days I'm working on Version 4.0.0, lots of backward-incompatible improvements in it (https://github.com/nikolaposa/version/blob/master/CHANGELOG.md#unreleased). 😒  BackwardCompatibilityCheck library is by far the biggest dependent so I'll help you with migrating, expect a PR soon!

File: src/Roave/SecurityAdvisories/Flag.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+final class Flag
+{
+    /**
+     * within extent of same version patch flag is of the highest priority
+     * e.g. 1.1-alpha < 1.1-beta < 1.1-rc < 1.1-stable < 1.1 < 1.1-p
+     */
+    private const PRIORITY = [
+        'patch'     => 5,
+        'p'         => 5,
+        null        => 4, // special case of clean version, e.g. 1.2.3
```
Comment (slash3b) [inline src/Roave/SecurityAdvisories/Flag.php]: I guess it looks a bit ugly, but when I added this `null` in the map, it helped to ditch so many useless checks.
So I think it is ugly and... elegant at the same time :sweat_smile:

File: src/Roave/SecurityAdvisories/Version.php [hunk 4]
```diff
@@ -24,25 +24,26 @@ final class Version
...
-        'p'         => 0,
-    ];
-
-    private function __construct()
+    /**
+     * @param string[] $matches
+     */
+    private function __construct(array $matches)
     {
+        $this->versionNumbers = self::removeTrailingZeroes(...array_map('intval', explode('.', $matches['version'])));
```
Comment (slash3b) [inline src/Roave/SecurityAdvisories/Version.php]: just moved this logic from `fromString` static method to unload it a little bit

File: src/Roave/SecurityAdvisories/Version.php [hunk 5]
```diff
@@ -57,47 +58,16 @@ public static function fromString(string $version) : self
...
-
-        return $object;
+        return new self($matches);
     }
 
     public function equalTo(self $other) : bool
     {
-        return $other->versionNumbers === $this->versionNumbers &&
-            $this->stabilityEqualTo($other);
+        return $other->versionNumbers === $this->versionNumbers
```
Comment (slash3b) [inline src/Roave/SecurityAdvisories/Version.php]: simplified logic, no need for `stabilityEqualTo`

File: test/RoaveTest/SecurityAdvisories/ComponentTest.php [hunk 6]
```diff
@@ -184,28 +184,79 @@ public function testSortAdvisoriesWithRealCase() : void
...
         ]);
 
         $component = new Component('foo/bar', $advisory1, $advisory2, $advisory3);
 
-        self::assertSame('>=2.1-beta.1,<2.1.3', $component->getConflictConstraint());
+        self::assertSame($expected, $component->getConflictConstraint());
+    }
+
+    /**
+     * @return mixed
```
[the remark this answers is not in the document]
Comment (slash3b) [reply test/RoaveTest/SecurityAdvisories/ComponentTest.php]: I did not know how to correct this :sweat: 
Is `@return mixed` okay ?

File: test/RoaveTest/SecurityAdvisories/VersionTest.php [hunk 7]
```diff
@@ -221,6 +221,21 @@ public function greaterThanComparisonVersionsProvider() : array
...
+            ['1.3.2-rc1', '1.3.2', false, true],
+            ['2.3.2-rc1', '2.3.2', false, true],
+            ['2.3.2', '2.3.3-alpha1', false, true],
+            ['2.3.2-p2', '2.3.2-alpha1', true, false],
+            // compare with version higher
+            ['2.3.2-p2', '2.3.3', false, true],
+            // compare with version lower
+            ['2.3.2-p2', '2.3.1', true, false],
+            // compare two patches but with additional stability versions like 1-p1 and 1-p2 so the p2 will be greater
+            ['2.3.2-p1', '2.3.2-p2', false, true],
```
Comment (slash3b) [inline test/RoaveTest/SecurityAdvisories/VersionTest.php]: @Ocramius I do this `p1` vs `p2` comparison here.
Also I have to admit it starts to get out of hand, data providers grow and it is a bit hard to navigate.

File: test/RoaveTest/SecurityAdvisories/VersionTest.php (hunk 1 above)
Comment (slash3b) [reply test/RoaveTest/SecurityAdvisories/VersionTest.php]: sorry, I deleted these lines, it was a work in progress

File: src/Roave/SecurityAdvisories/Version.php (hunk 2 above)
Comment (slash3b) [reply src/Roave/SecurityAdvisories/Version.php]: threw it away.  
Hope your eyes are okay = )

**Comment by slash3b** [comment]
It is ready for review :crossed_fingers:

File: src/Roave/SecurityAdvisories/Flag.php (hunk 3 above)
Comment (stof) [reply src/Roave/SecurityAdvisories/Flag.php]: `null` is not a valid array key. This is probably adding the empty string instead.

Comment (slash3b) [reply src/Roave/SecurityAdvisories/Flag.php]: :thinking: I'd say it actually is a valid key, at least I'm being able to get value from array by `null`   

But you got me thinking, and yes `null` is not very good key. I'm going to change it to string value with some meaning attached.   

Thanks!

Comment (slash3b) [reply src/Roave/SecurityAdvisories/Flag.php]: In the end empty string was the best variant. I did push update.

File: test/RoaveTest/SecurityAdvisories/ComponentTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ocramius) [reply test/RoaveTest/SecurityAdvisories/ComponentTest.php]: It's a test data provider: we're good for now 👍

File: test/RoaveTest/SecurityAdvisories/VersionTest.php (hunk 7 above)
Comment (Ocramius) [reply test/RoaveTest/SecurityAdvisories/VersionTest.php]: Yeah, but that's combinatory logic with edge cases 🤷‍♀️

The good thing is that I'm not (currently) aware of further edge cases

**Comment by slash3b** [comment]
@Ocramius thank you!    
 I'm so glad I could help ^^

**Comment by Ocramius** [comment]
Seems fixed downstream :dancer: https://github.com/Roave/SecurityAdvisories/commit/72ffba7ecbc58c109000bff92af4ee5db8fb1208#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780R78

**Comment by nikolaposa** [comment]
@slash3b What does `-p` actually denote? Is that some custom identifier? It definitely isn't [Semver](https://semver.org/)-compliant because the correct order should be:

`1.1-alpha < 1.1-beta < 1.1-p < 1.1-rc < 1.1-stable < 1.1`

> Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows: identifiers consisting of only digits are compared numerically and **identifiers with letters or hyphens are compared lexically in ASCII sort order**.

**Comment by Ocramius** [comment]
`-p` is for `-patch`, so `1.2.0-p1` is a bit higher than `1.2.0`. I was as confused as you when Magento decided to release with this modifier 🤷‍♀️

**Comment by nikolaposa** [comment]
Oh no.
